src/solver/cumulative_reduction/forward_cumulative_reduction_contiguous_lastdim.cpp

```diff
@@ -0,0 +1,146 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/datatype.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/cumulative_reduction/invoke_params.hpp>
+#include <miopen/cumulative_reduction/solvers.hpp>
+#include <miopen/cumulative_reduction/utils.hpp>
+
+#define FLOAT_ACCUM float
+
+#define LOCAL_SIZE_MAX 256
+#define LOCAL_SIZE_MIN 64
+
+namespace miopen {
+
+namespace solver {
+
+namespace cumulative_reduction {
+
+bool IsImprovementOverROCm(const ExecutionContext& /*context*/,
+                           const miopen::cumulative_reduction::ForwardProblemDescription& problem)
+{
+    if(problem.GetInputDesc().GetLengths()[problem.GetDim()] > LOCAL_SIZE_MAX)
+        return false;
+    return true;
+}
+
+bool ForwardContiguousLastDim::IsApplicable(
+    const ExecutionContext& context,
+    const miopen::cumulative_reduction::ForwardProblemDescription& problem) const
+{
+    if(!IsImprovementOverROCm(context, problem))
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    if(problem.GetInputDesc().GetStrides()[problem.GetDim()] != 1)
+        return false;
+    if(problem.GetInputDesc().GetSize() == problem.GetOutputDesc().GetSize() &&
+       problem.GetOutputDesc().GetStrides()[problem.GetDim()] != 1)
+        return false;
+    if(problem.GetInputDesc().GetSize() == problem.GetIndicesDesc().GetSize() &&
+       problem.GetIndicesDesc().GetStrides()[problem.GetDim()] != 1)
```

Comment:
Even if the indices tensor is contiguous, this condition will likely be false unless dim is the last dimension of the tensor. Is this the intended behavior?

Reply:
Kernel name is ContiguousLastDim, it seems to be correct.